### DISCUSSION TITLE
How to convert GIF to web-friendly MP4 (MPEG-4) and WEBM?

### DISCUSSION THREAD

**Discussion by tmb-github** [ImageMagick/ImageMagick#5485] [Help] [answered]
I'm trying to convert GIF format files to MP4 (MPEG-4) and WEBM using ImageMagick. I want to optimize them (file size, image quality) for display on the web.

There are many examples of going from MP4 to GIF that I've found, but the few I've found for GIF to MP4 involve first splitting the GIF into a series of PNGs and then combining them into a GIF using FFMPEG.

I'm seeking a one-step solution using ImageMagick alone, including the appropriate codec(s), etc.

Many thanks!

**Comment by snibgo** [comment, marked as answer]
The obvious commands are:
```
magick in.gif out.mp4

magick in.gif out.webm
```
If your GIFs may be compressed, use "-layers coalesce" after reading the GIF. 

IM uses ffmpeg to write the outputs. Use "-verbose" to see the ffmpeg command used. IM does not give full access to the ffmpeg options. If you need that, then run ffmpeg directly.

**Reply by fmw42** [reply]
You could customize the delegates.xml file to add other ffmpeg options if you need them.

**Reply by tmb-github** [reply]
@snibgo  The MP4s and webms are converting fine now that ffmpeg is in place, and the output files play in Chrome without a problem. But Firefox cannot decode them. Are there certain conversion settings that should be used to ensure they can run in Firefox?

**Reply by tmb-github** [reply]
@fmw42 I may need to do that to ensure the converted files can be played in Firefox. Currently, Firefox is not playing the MP4s, WEBMs, or OGVs I'm making. Apparently, I need to ensure the MP4 conversions are H.264 (Mozilla states Firefox should play MP4s with that encoding). What would I add to the delegates.xml file to allow this?

**Reply by tmb-github** [reply]
@fmw42 Update: The conversion that works for Firefox is:

ffmpeg -i input.gif -movflags faststart -pix_fmt yuv420p -vf "scale=trunc(iw/2)*2:trunc(ih/2)*2" output.mp4

How would this be stored in the delegates.xml file?

**Reply by fmw42** [reply]
Sorry, I am not an expert on editing the delegates.xml file.

**Comment by tmb-github** [comment]
@snibgo Running the command you suggest produces the following error message:

> magick: VideoDelegateFailed `Unrecognized option 'plays'.
> Error splitting the argument list: Option not found
> ' @ error/video.c/WriteVIDEOImage/698.

**Comment by snibgo** [comment]
> VideoDelegateFailed

This suggests that IM can't find ffmpeg.

When I install IM on Windows, it asks if I want to include ffmpeg. I always say "yes".

**Comment by snibgo** [comment]
In an IM command, you can include...
```
-define video:pixel-format=yuv420p
```
... to get that pixel format.

You can edit a line in delegates.xml to include other desired options. For example:
```
  <delegate decode="gif" encode="video:encode" stealth="True" command="&quot;ffmpeg.exe&quot; -nostdin -loglevel error -i &quot;%s%%d.%s&quot; %s -movflags faststart -vf &quot;scale=trunc(iw/2)*2:trunc(ih/2)*2&quot; &quot;%s.%s&quot;"/>
```

**Reply by tmb-github** [reply]
@snibgo Many thanks! This worked after one tweak (it failed initially with error messages). There seems to be one mistake in the delegate code.:

`&quot;%s%%d.%s&quot;`

...should perhaps be:

`&quot;%s%%d.pam&quot;`

I studied the other delegate codes in the xml file, and 'pam' was used where the 2nd %s is located in the snippet above. Once I made this change, it worked.

Many thanks!

**Reply by swistak** [reply]
Thanks for the advice, however I tried adding `-define video:pixel-format=yuv420p` in different places of:

```
convert -delay 10 'inter/*.png' -define video:pixel-format=yuv420p -loop 0 -morph 5 results/test.mp4
```

and it gets ignored, any advice where should it go?

Version: ImageMagick 6.9.11-60 Q16 x86_64 2021-01-25 https://imagemagick.org

**Reply by fmw42** [reply]
That define may only be available for current versions of IM and perhaps requires current version of IM 7. I do not know when that define was introduced.  Can you upgrade to the latest IM 7?

**Reply by swistak** [reply]
I checked in the meantime and indeed it requires `7.1` for that option. Distro I'm using only has 6.9 in it though, might need to get extra package sources. 

Thanks for help!

**Reply by fmw42** [reply]
Install your own separate version. You can install the portable version (magick) if you want.  See magick at https://imagemagick.org/script/download.php

**Comment by snibgo** [comment]
We may have different versions of IM. I am using 7.1.0-42. I simply inserted your required options into an existing line, without changing it in any other way. You should do the same.

For video MP4 and similar (processed by `coders/video.c`), translations of the parameters "%s" etc are hard-coded in IM, so we can't simply add or subtract them from the delegate line. An alternative is to add a new format with any name we like, such as "mymp4", like this:
```
  <delegate decode="gif" encode="mymp4" command="&quot;ffmpeg.exe&quot; -nostdin -loglevel error -i &quot;%i&quot; -movflags faststart -pix_fmt yuv420p -vf &quot;scale=trunc(iw/2)*2:trunc(ih/2)*2&quot; -f mp4 -y &quot;%o&quot;"/>
```

This isn't a "video" format, so escapes %i, %o etc are as documented at the top of delegates.xml. I have included ffmpeg options "-f mp4" to force the output format, and "-y" to overwrite the output file if it already exists. You might want to include "-y" in the delegate line I previously showed.

We can use it like this:
```
magick in.gif -layers coalesce out.mymp4
```
This has made a file `out.mymp4`, which is a valid MP4 file, but should probably be renamed `out.mp4`.

As a general rule, when I *add* lines to delegates.xml, I put them in my local file, not the generic one. This means my additions survive IM upgrades. On my Windows 8.1 laptop, the local delegates.xml is `C:\Users\Alan\AppData\Local\ImageMagick\delegates.xml`
